### PR TITLE
completed week 7 assessment

### PR DESCRIPTION
**Pull request by MikeSC27**
# Submitting Week 7 Assessment

### ✅ Completed Files

Select the following as confirmation that the files in this repository are complete.

- [X] Coding Challenges
- [X] Coding Concepts
- [X] Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:
As much as this assessment was challenging, it was fun.  It reminded me to keep it simple sometimes.  The user stories were more fun than I first thought.  Please let me know if you think I got off track on its purpose.  Thank you.

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -44,3 +280,30 @@ const cubeAndSum2 = [0, 5, 10]
...
+const sumCubed = (array) => {
+  let cube = array.map((value) => value * value * value)
+  let total = 0
+  for(let number of cube) {
+    total += number;
+  }
+  return total
+}
+console.log(sumCubed(cubeAndSum1))
+console.log(sumCubed(cubeAndSum2))
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:301]: Very good! There is a tool called reduce that would be very helpful here.

File: code-challenges.test.js [hunk 2]
```diff
@@ -10,32 +14,264 @@
...
+        //Return the numbers after modulo operation. In retrospec this was much easier than it was for me becaues I got stuck on the isNaN function and Number.isNaN method.  If I just thought to seperate into two arrays first I would have been done as fast as I finished problem 3.  
+
+const threeRemain = (array) => {
+    let numbs = []
+    let bad = []
+    array.forEach((element) => typeof element === 'number' ? numbs.push(element) : bad.push(element));
+      return numbs.map((number) => {
+      return number % 3
+    }) 
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:192]: So why use forEach and push if map does the same thing?

File: code-challenges.test.js [hunk 3]
```diff
@@ -10,32 +14,264 @@
...
+        let name = (Object.values(array).map((value)=>value.name.split(" ").map(item=> item.charAt(0).toUpperCase() + item.substr(1))))
+            let one = name[0].join(' ')
+            let one1 = work[0]
+            let two = name[1].join(' ')
+            let two2 = work[1]
+            let three = name[2].join(' ')
+            let three3 = work[2]
+        return [`${one} is ${one1}.`, `${two} is ${two2}.`, `${three} is ${three3}.`]
+        }
+ console.log(aboutMe(people))
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:72]: Thi sgets the job done but there are tools that would allow this process to be more dynamic and allow for use cases that are not a part of these testing variables. ALWAYS consider the edge cases.

File: interview-questions.md [hunk 4]
```diff
@@ -8,36 +8,86 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+Researched answer: Creating the migration file for the change is a good place to start here; $rails g migration add_foreign_key_to_students
+Then make sure in the method that uniqueness is true on the Cohort table and add foreign key to Student table.  Dont forget to $rails db:migrate.
 
 2. Which RESTful routes must always be passed params? Why?
 
-Your answer:
+Your answer: Well using params with the routes makes the overall app or web page more "dynamic".
 
-Researched answer:
+Researched answer:  Using params helps to not only make it overall more "Dynamic" but also lets you have more information, just create the method associated with your value.  Using params will help not only dynamically but reduce your over all scripting and perhaps the efficiency as well since it is data being sent back from the client to the servers, (sometimes in Idaho).
```
Comment (WalkerTexas-coder) [inline interview-questions.md:20]: This isn't wrong but it does not answer the question of which RESTful routes require params.

File: interview-questions.md [hunk 5]
```diff
@@ -8,36 +8,86 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 3. Name three rails generator commands. What is created by each?
 
-Your answer:
+Your answer: 
+$ rails g controller XxxxXxx - this generator command creates your main home controller for your app currently called XxxxXxx, which is in PascalCase singular. It not only makes the controller file, it also creates the veiw file, and a functional test file, a javascript file, and css type file. 
 
-Researched answer:
+$ rails g model Table name:string shoe_size:integer -  this generator creates the model for the table called Table with columns name and shoe_size.  It aslo states the allowed datatype to be inputed into each coloumn. 
+
+$ rails g rspec:install - This installs the dependancies for rspec which should be done before generating the model.  Doing it in this order installs a directory called spec with a model folder automatically when generating the model.  This is all for our TDD and validations in the app.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:29]: This is great!

File: interview-questions.md [hunk 6]
```diff
@@ -8,36 +8,86 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+As a user who is busy I need to be able to adjust my priorities according to my change in schedule. Also I need to be able to adjust the color code. 
+
+User Story 9:
+As a user who is busy I need to have a notification of my changes saved and sent to my email for update purposes to reflect all changes.
+
+User Story 10:
+As a user who is busy I need to be able to close out tasks as complete with log of time of completion.
+
+User Story 11:
+As a user who is busy I need to be notified of all tasks completed ahead of time and be given an "ATTA BOY" for a morale boost.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:93]: This is well done. "Who is busy" is very silly.

File: rails-commenting.rb [hunk 7]
```diff
@@ -49,14 +57,17 @@ def destroy
     @post = BlogPost.find(params[:id])
     if @post.destroy
       # ---8)
+  # This is the route to delete the specific blog post id from the database.
       redirect_to blog_posts_path
     end
   end
 
   # ---9)
+# This is the super secret extra tough Strong Params, for extra security, keeping would be bad actors out of your routes and controllers to protect your database unless specifically allowed. This is meant to prevent Mass Assignment where a bad actor could gain privileges such as admin.  This is basically assigning data to mulitple database sections at the same time which could lead to allowing them access to overwrite data not meant to be changed.
```
Comment (WalkerTexas-coder) [inline rails-commenting.rb:66]: This is referring to the word private a not the blog post params. 

You are on the right path for the strong params though!

**Review by WalkerTexas-coder — APPROVED** [review]
Mike this is a solid assessment. I am marking this as Competent, but I want to encourage you to research making your code more dynamic and less static. Your interview answers are solid and your rails concepts are headed in the right direction but if you find the time review these concepts and perhaps talk them over with someone to help them solidify. You are all done! Approved!
